### ISSUE TITLE
Default implementations of methods in Tables.jl

### ISSUE THREAD

**Issue by bkamins** [JuliaData/Tables.jl#5]
I started to test the package and I have one thing that was not clear to me.
Why do you provide default implementations for the methods in this package, e.g. `rows` and `columns`, `schema`. Users should implement a specific method for them anyhow if they are to be used so maybe it would be better if by default they thrown "not imlemented" error?

(I am starting to digest the package so I might be missing something)

**Comment by quinnj** [comment]
Take the following scenario:

* CSV.File naturally returns a Row-iterator, and thus implements `Tables.rows`, the effort to support `Tables.columns` is non-trivial; you'd have to drastically change the internal data structure, or find some other means of eagerly reading an entire file and internally store the full columns to properly provide a `Tables.columns` implementation
* DataFrames, in its sink constructor, however, has a very easy definition like `DataFrame(x::Any) = DataFrame([collect(u) for u in Tables.columns(x)], collect(Tables.names(Tables.schema(x))))`. I.e. it's super easy to construct a DataFrame if you have the result of `Tables.columns`. Similarly, it would be non-trivial to implement a DataFrames sink method that only accepted `Tables.rows`; it would have to allocate columns, loop over rows, either push or setindex depending on if the `Tables.rows` iterator knew its length or not

The nice thing is that the two generic fallback implementations are pretty efficient and general for table types: i.e. the default `Tables.rows` for a table that implements `Tables.columns` is a non-allocating RowIterator that provides direct views into the original columns. So instead of requiring a columnar table to provide it's own `Tables.rows`, it can just use the default implementation and it will be efficient.

If a table implements `Tables.rows`, the default `Tables.columns` eagerly builds up columns to return. It's convenient because it's essentially the same work a package like CSV.jl would have to implement if it tried to do `Tables.columns` itself.

Hope that helps clarify.

**Comment by bkamins** [comment]
I understand the idea of `rows` and `columns` duality, what I am not clear about is why the package assumes that by default `AccessStyle(x) = RowAccess()` and its consequences. What I would normally expect is e.g.:

* `AccessStyle(x) = nothing` by default
* `schema(x) = throw(DomainError("$x does not support Tables interface"))  # has to be explicitly implemented`
* in `rows(x::T)` initially make a check `AccessStyle(x) === nothing` and throw error on success
* similarly in `columns(x::T).

or something similar. In general is not clear for me why `RowAccess()` is the default.

Of course then a special functions for vector of namedtuples and namedtuple of vectors might be needed but it should be simple.

**Comment by quinnj** [comment]
It's really just for convenience. It also happens to simplify the integration w/ queryverse because `Enumerable`s implicitly satisfy the `Tables.rows` side of the interface w/ the defaults. I guess it would be pretty simple to implement those in the Requires-ed `enumerable.jl` file, but I don't know, I don't really see much harm in choosing a default there.

**Comment by quinnj** [comment]
I'm certainly open to the definitions you proposed; I'll have to think about it. The current defaults, like I said, were really just because it simplified some of the initial implementations.

**Comment by bkamins** [comment]
The only harm is debugging. E.g. I have just commented against `DataFrames(x)` catch-all method because of this reason as when you run it on an object that does not support Tables.jl interface you get an obscure error message that most users will not understand.

OTOH I think it is enough to add `AccessStyle` to `enumerable.jl` and all should work (even `schema` could have a default implementation `schema(x) = AccessStyle(x) === nothing ? error("something") : eltype(x)` if you find it convenient).

And default `AccessStyle` could be `nothing` or something like `struct UndefinedAccess <: AccessStyle end` to keep type stability (It should not make a huge difference in Julia 1.0).

Also maybe default methods should explicitly check for `RowAccess` and `ColumnAccess` as someone might create yet another subtype of `AccessStyle` (not likely but who knows what future would bring, e.g. something like `BatchRowAccess` indicating that it is most efficient to process the table in chunks as it is e.g. distributed like in JuliaDB).

In short - I really like the design so - hopefully - if will become a standard. If it becomes a standard then API should as be newbie friendly (read: safeguarding them against their own mistakes) as possible.

**Comment by nalimilan** [comment]
I'd say it's fine to default to `RowAccess` (just like Base defaults to `IndexLinear`), but I also think it would be nice to have a trait to avoid trying to treat any argument as a table (just like Base uses `AbstractArray`), so that we don't print obscure errors to users.

**Comment by bkamins** [comment]
It is OK for me to have a second trait indicating Tables.jl support.

**Comment by mdavezac** [comment]
👍 for a trait. Catch-all functions invite too much vendor (a.k.a. Tables.jl in this case) lock-in.
